packages/@aws-cdk/aws-codepipeline/lib/action.ts

```diff
@@ -111,20 +226,272 @@ export interface ActionBindOptions {
+   */
+  public get actionProperties(): ActionProperties {
+    return {
+      ...this._actionProperties,
+      // Implied defaults here
+      owner: this._actionProperties.owner ?? 'AWS',
+      version: this._actionProperties.version ?? '1',
+      // Replace supplied value with a token
+      variablesNamespace: this.namespaceOrToken,
+      // Replace potentially mutated values with their final values
+      runOrder: this.runOrder,
+      actionName: this.actionName,
+      inputs: this.inputs,
+      outputs: this.outputs,
+    };
+  }
+
+  /**
+   * The name of this Action
+   */
+  public get actionName(): string {
+    return this._actionName;
+  }
+
+  /**
+   * The name of this Action
+   */
+  public set actionName(value: string) {
+    validation.validateName('Action', value);
+    this._actionName = value;
+  }
+
+  /**
+   * The runOrder of this Action
+   */
+  public get runOrder(): number {
+    return this._runOrder;
+  }
+
+  /**
+   * The runOrder of this Action
+   */
+  public set runOrder(value: number) {
+    validateRunOrder(value);
+    this._runOrder = value;
+  }
+
+  /**
+   * Inputs to this action
+   */
+  public get inputs(): Artifact[] {
+    return deduplicateArtifacts(this._inputs);
+  }
+
+  /**
+   * Ouptuts of this action
+   */
+  public get outputs(): Artifact[] {
+    return deduplicateArtifacts(this._outputs);
+  }
+
+  public bind(scope: Construct, stage: IStage, options: ActionBindOptions): ActionConfig {
+    this._pipeline = stage.pipeline;
+    this._stage = stage;
+    this._scope = scope;
+
+    // default a namespace name, based on the stage and action names
+    this.actualNamespace = this.customerProvidedNamespace ?? `${stage.stageName}_${this.actionProperties.actionName}_NS`;
+
+    // For backwards compatibility reasons, the result of 'bound' is
+    // the default configuration. Subclasses can override `configuration` to return something else.
+    return this.bound(scope, stage, options);
+  }
+
+  public onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps) {
+    const rule = new events.Rule(this.scope, name, options);
+    rule.addTarget(target);
+    rule.addEventPattern({
+      detailType: ['CodePipeline Action Execution State Change'],
+      source: ['aws.codepipeline'],
+      resources: [this.pipeline.pipelineArn],
+      detail: {
+        stage: [this.stage.stageName],
+        action: [this.actionProperties.actionName],
+      },
+    });
+    return rule;
+  }
+
+  /**
+   * Change inputs to this action
+   *
+   * Protected so that subclasses can exert control over how many artifacts there
+   * are and how they are replaced.
+   */
+  protected changeInputs(values: Artifact[]) {
+    this._inputs = [...values];
+  }
+
+  /**
+   * Outputs of this action
+   *
+   * Protected so that subclasses can exert control over how many artifacts there
+   * are and how they are replaced.
+   */
+  protected changeOutputs(values: Artifact[]) {
```

Comment:
Disagree twice.

I feel like the `inputs`/`outputs` need to be protected because it's the concrete class that decides whether there are 0 outputs, 1 output, N outputs. Having arbitrary users be able to mutate outputs willy nilly is almost guaranteed to destroy correct  operation of the class.

Subclasses will decide whether and how to expose this mutator.